### PR TITLE
Add modeling and utils changes for transformers=v4.41.2

### PR DESCRIPTION
**Pull request by vbaddi**
- Update transformers version from `v4.36.2 `to `v4.41.2`
- Add range-based gather and scatter custom ops.
- Update `Llama `, `Mistral ` and `Mixtral `Modeling changes with range-based gather/scatter.
- Export and latency script changes to include new set of inputs. *remove `attention_mask ` from inputs.
- `Onnx `and `Onnx-Runtime` version updates
- `todo`: Add changes to all other model categories and api testing

### CODE REVIEW

**Comment by quic-mamta** [comment]
Need to update padding side to right in generate_inputs.py and run_utils.py also

File: QEfficient/cloud/infer.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def main(
         ignore_patterns=["*.txt", "*.onnx", "*.ot", "*.md", "*.tflite", "*.pdf", "*.msgpack", "*.h5"],
     )
     tokenizer = AutoTokenizer.from_pretrained(
-        model_hf_path, use_cache=True, padding_side="left", trust_remote_code=True
+        model_hf_path, use_cache=True, padding_side="right", trust_remote_code=True
```
Comment (irajagop) [inline QEfficient/cloud/infer.py:70]: Looks like `trust_remote_code=True` is causing issues for some models, e.g: Phi2, Phi3. I think we should remove that or make it a flag whether to use remote code or not.

File: QEfficient/cloud/infer.py [hunk 2]
```diff
@@ -115,7 +115,7 @@ def main(
     #############################################
     # hf model -> export -> compile -> execute
     #############################################
-    model_hf = AutoModelForCausalLM.from_pretrained(model_hf_path, use_cache=True)
+    model_hf = AutoModelForCausalLM.from_pretrained(model_hf_path, use_cache=True, trust_remote_code=True, attn_implementation="eager")
```
Comment (irajagop) [inline QEfficient/cloud/infer.py:118]: Here also, we should either remove the `trust_remote_code=True` or make it depend on a flag.

File: QEfficient/cloud/infer.py (hunk 1 above)
Comment (vbaddi) [reply QEfficient/cloud/infer.py]: Sure, addressed this. In a larger change, Will make it part of `kwargs` as an optional flag to enable custom models, tokenizers.

File: QEfficient/cloud/infer.py (hunk 2 above)
Comment (vbaddi) [reply QEfficient/cloud/infer.py]: Done

**Comment by vbaddi** [comment]
**Latest updates: 2024_06_07**
- Update to `transformers`, `onnx`, `onnxruntime `versions. (see `setup.py` for more details)
- Add `range-based` `gather `and `scatter `custom ops in the pipeline.
- Models updated/added with the new interface `LLama`, `Mistral`, `Mixtral`, `CodeGen`, `Phi3`, `Qwen2`, `Starcoder2`. 
- Move the `attention_mask `generation using position ids to the common utils.
- Export and latency script changes to include new set of inputs. *remove `attention_mask ` from inputs.
- Changes to `generate_inputs `and `run_utils `to adhere to the new interface changes.
- Tests config modification.
- `todo`: models pending for update `GPT`, `MPT`

**Comment by anujgupt-github** [comment]
- We should add details in readme also as part of this PR, the changes to the inputs (as a note) and arguments of LL APIs
- As part of this PR, you are bringing/merging support of phi-3 also, so that should also come in Readme
- More review WIP

File: QEfficient/generation/text_generation_inference.py [hunk 3]
```diff
@@ -145,108 +145,98 @@ def cloud_ai_100_exec_kv_helper(
...
-    if generation_len is None:
-        generation_len = ctx_len
-    num_chunks = -(input_len // -prompt_len)  # ceil divide without float
-    input_len = num_chunks * prompt_len  # Convert input_len to a multiple of prompt_len
-    assert input_len <= ctx_len, "input_len should be less than ctx_len"
-
     # Skip inputs/outputs
-    session.skip_buffers([x for x in session.input_names if x.startswith("past_")])
-    session.skip_buffers([x for x in session.output_names if x.endswith("_RetainedState")])
+    session.skip_buffers([x for x in session.input_names + session.output_names if x.startswith("past_")])
```
Comment (anujgupt-github) [inline QEfficient/generation/text_generation_inference.py:158]: Why this change? are we appending past_ to attention mask also?

File: QEfficient/transformers/modeling_utils.py [hunk 4]
```diff
@@ -11,70 +11,75 @@
...
-    MixtralBLockSparseTop2MLP,
-    MixtralDecoderLayer,
     MixtralForCausalLM,
     MixtralModel,
     MixtralRMSNorm,
-    MixtralRotaryEmbedding,
     MixtralSparseMoeBlock,
 )
 from transformers.models.mpt.modeling_mpt import MptAttention, MptBlock, MptForCausalLM, MptModel
+from transformers.models.phi3.modeling_phi3 import Phi3Attention, Phi3ForCausalLM, Phi3Model, Phi3RMSNorm
```
Comment (anujgupt-github) [inline QEfficient/transformers/modeling_utils.py:38]: can we follow uniform style? 
same line for all modules or separated in new lines?

File: tests/config.json [hunk 5]
```diff
@@ -4,17 +4,25 @@
...
+    {
+      "model_name": "tiiuae/falcon-7b",
+      "model_class": "FalconForCausalLM"
+    },
+    {
+      "model_name": "Qwen/Qwen2-0.5B",
+      "model_class": "Qwen2ForCausalLM"
+    },
+    {
+      "model_name": "bigcode/starcoder2-3b",
```
Comment (anujgupt-github) [inline tests/config.json:24]: can we add one for starcoder 15,5B?

**Comment by vbaddi** [comment]
> * We should add details in readme also as part of this PR, the changes to the inputs (as a note) and arguments of LL APIs
> * As part of this PR, you are bringing/merging support of phi-3 also, so that should also come in Readme
> * More review WIP

Sure, will update readme.

File: QEfficient/generation/text_generation_inference.py (hunk 3 above)
Comment (vbaddi) [reply QEfficient/generation/text_generation_inference.py:158]: no, just skipping the retention `past_kvs `in the qpc execution.

File: QEfficient/transformers/modeling_utils.py (hunk 4 above)
Comment (vbaddi) [reply QEfficient/transformers/modeling_utils.py:38]: Hmm, this is coming from `linter `checks, post the 100 lines, it will take it to next line.

File: tests/config.json (hunk 5 above)
Comment (vbaddi) [reply tests/config.json:28]: The idea was to test the pipeline with smaller model checkpoints to get the output and execution correctness. 
I can add one larger set for starcoder2. Maybe. [bigcode/starcoder2-7b](https://huggingface.co/bigcode/starcoder2-7b), is that fine?

File: QEfficient/exporter/export_utils.py [hunk 6]
```diff
@@ -17,6 +17,24 @@
...
+#     except Exception as e:
+#         print(f"Skipping simplifier: {e}")
+#         onnx.shape_inference.infer_shapes_path(
+#             f"{gen_models_path}/{model_base_name}.onnx",
+#             f"{gen_models_path}/{model_base_name}_inf.onnx",
+#             True,
+#             True,
+#             True,
+#         )
+#         return save_onnx(model_base_name + "_inf", gen_models_path, model_base_name)
```
Comment (irajagop) [inline QEfficient/exporter/export_utils.py:37]: We can just remove this and create a function for shape_inference only, as onnxsim creates more issues that it solves. Can be done in a future PR too.

File: QEfficient/generation/text_generation_inference.py (hunk 3 above)
Comment (irajagop) [reply QEfficient/generation/text_generation_inference.py:158]: @anujgupt-github There is no attention_mask retention now, so no need of separate skipping logic for inputs and outputs.

File: QEfficient/transformers/modeling_attn_mask_utils.py [hunk 7]
```diff
@@ -12,6 +12,43 @@
 from transformers.modeling_attn_mask_utils import AttentionMaskConverter
 
 
+def _update_causal_mask(
```
Comment (irajagop) [inline QEfficient/transformers/modeling_attn_mask_utils.py:15]: Should we name this better? We're not "updating" the causal_mask here. May be `_create_causal_mask`?

**Review by irajagop — CHANGES_REQUESTED** [review]
Only the last comment needs to be addressed here. There are other clean-up comments apart from these (not relevant here), I'll provide those after merging this PR.

File: QEfficient/transformers/modeling_attn_mask_utils.py (hunk 7 above)
Comment (vbaddi) [reply QEfficient/transformers/modeling_attn_mask_utils.py]: yes, makes sense, Updated.

**Review by irajagop — APPROVED** [review]
